### PR TITLE
[Form] DateType storing wrong format date

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #24791
| License       | MIT
| Doc PR        | 

I don't think adding the logger as a dependencies to the form is a good idea, how could we log without it directly and without using trigger_error ?

EDIT: throw added.

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/Type/DateType.php
Comment (fabpot) [inline src/Symfony/Component/Form/Extension/Core/Type/DateType.php]: I would throw an `InvalidOptionsException` instead.

Comment (Simperfit) [reply src/Symfony/Component/Form/Extension/Core/Type/DateType.php]: Ill do that.

**Comment by Simperfit** [comment]
AppVeyoy failure is unrelated and @fabpot's comments fixed.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/DateTypeTest.php
Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/DateTypeTest.php]: let's use the `@expectedException` annotation instead

Comment (Simperfit) [reply src/Symfony/Component/Form/Tests/Extension/Core/Type/DateTypeTest.php]: done

**Comment by Simperfit** [comment]
This PR seems ready for review.

**Comment by xabbuh** [comment]
2.7 and 2.8 are not affected?

**Comment by Simperfit** [comment]
I think they are, PR rebased

File: src/Symfony/Component/Form/Extension/Core/Type/DateType.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/DateType.php]: This forces the yMd letters to always be all there - but the check on L54-58 below allows only one of them.
Is this really legitimate?

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/DateTypeTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/DateTypeTest.php]: Does this really need to change (same below?) Is it really legitimate?

Comment (Simperfit) [reply src/Symfony/Component/Form/Tests/Extension/Core/Type/DateTypeTest.php]: I think it is because if you pass 'yy', the message is wrong, IMO.

Comment (Simperfit) [reply src/Symfony/Component/Form/Tests/Extension/Core/Type/DateTypeTest.php]: This is right, yy is not valid in ICU.

File: src/Symfony/Component/Form/Extension/Core/Type/DateType.php
Comment (Simperfit) [reply src/Symfony/Component/Form/Extension/Core/Type/DateType.php]: You are right, I've forgot something the Regexp has been updated.

**Comment by Simperfit** [comment]
@nicolas-grekas could you please look at this a second time ? It feels rdy to me.

**Comment by Simperfit** [comment]
This one still needs a second review.

**Comment by Simperfit** [comment]
ping @HeahDude if you have some time to review this. it's ready ;).

**Review by ostrolucky — CHANGES_REQUESTED** [review]
- Needs more investigation in issue. Doesn't seem possible to experience it
- This fix is BC break
- This fix is conceptually wrong, as it prevents using e.g. "y" for single_text

**Pull request closed without merging**
